File: pset6/crack.py
import cs50
import sys
import crypt

MAX_PASS_LENGTH = 4
PASSWORD_CHARS  = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

def main():
    if len(sys.argv) != 2:
        print("Usage: ./crack hash")
        sys.exit(1)

    hash_string = sys.argv[1]
    for size in range(1, MAX_PASS_LENGTH + 1):
        outPassword = bruteForce(hash_string, size)
        if outPassword != None:
            print(outPassword)
            sys.exit(0)
    sys.exit(-1)


def bruteForce(hash_string, charsCount):
    hash_len = len(hash_string)
    passwordCharsLen = len(PASSWORD_CHARS)

    optionsCount = 1
    for i in range(0, charsCount):
        optionsCount *= passwordCharsLen

    salt = hash_string[0] + hash_string[1]
    
    for i in range(optionsCount):
        index = i
        password = ""
        for j in range(charsCount):
            password += PASSWORD_CHARS[index % passwordCharsLen]
            index //= passwordCharsLen

        #print(" -> {}".format(password))
        if crypt.crypt(password, salt) == hash_string:
             return password

    return None


if __name__ == "__main__":
    main()

File: pset2/crack/crack.c
#define _XOPEN_SOURCE

#include <unistd.h>
#include <stdio.h>
#include <cs50.h>
#include <string.h>

#define MAX_PASS_LENGTH 4

const string passwordChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

string bruteForce(string hash, int charsCount);

int main(int argc, string argv[]) {

    if (argc != 2) {
        printf("Usage: ./crack hash\n");
        return 1;
    }

    string hash = argv[1];
    for (int size = 1; size <= MAX_PASS_LENGTH; size++) {
        string outPassword = bruteForce(hash, size);
        if (outPassword != NULL) {
            printf("%s\n", outPassword);
            return 0;
        }
    }
    
    return -1;
}

bool strEqual(unsigned long len, string s1, string s2) {
    for(int i = 0; i < len; i++) {
        if (s1[i] != s2[i])
            return false;
    }
    return true;
}

string bruteForce(string hash, int charsCount) {
    unsigned long len = strlen(hash);
    unsigned long passwordCharsLen = strlen(passwordChars);

    long long optionsCount = 1;
    for (int i = 0; i < charsCount; i++) {
        optionsCount *= passwordCharsLen;
    }

    char *password = malloc((charsCount + 1) * sizeof(char));
    password[charsCount] = '\0';
    
    char salt[3];
    salt[0] = hash[0];
    salt[1] = hash[1];
    salt[2] = '\0';

    for (long long i = 0; i < optionsCount; i++) {
        int index = i;
        for(int j = 0; j < charsCount; j++) {
            password[j] = passwordChars[index % passwordCharsLen];
            index /= passwordCharsLen;
        }
        
        //printf(" -> %s\n", password);
        string thisHash = crypt(password, salt);
        if (strEqual(len, hash, thisHash))
             return password;
    }

    free(password);
    return NULL;
}